Interactivity API: implement the new `store()` API

DAreRodz opened the pull request:
## What

This PR implements the new `store()` API as specified in the proposal: https://github.com/WordPress/gutenberg/discussions/53586

It also updates the current Interactivity API implementation to match the latest features and improvements developed in [WordPress/gutenberg](https://github.com/WordPress/gutenberg/).

Will close https://github.com/woocommerce/woocommerce-blocks/issues/11065

## Why

To test the new API with actual use cases so we can fine-tune and shape it.

## Testing Instructions

Please, go to the following PRs that are built on top of this branch to test these changes:

- https://github.com/woocommerce/woocommerce-blocks/pull/11558
- https://github.com/woocommerce/woocommerce-blocks/pull/11721

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

luisherranz committed:
Sync Interactivity API code with Gutenberg

luisherranz committed:
New store() API

luisherranz committed:
Store raw actions

DAreRodz committed:
Update wc-interactivity-store implementation

DAreRodz committed:
Replace `wc_store` with `wc_initial_state`

DAreRodz committed:
Parse and populate initial state

DAreRodz committed:
Allow store parts in `store()`

DAreRodz committed:
Accept namespaces in directive paths

DAreRodz committed:
Add $$namespace to directives' object values

DAreRodz committed:
Make namespace parsing more robust

DAreRodz committed:
Use DeepPartial type for store parts

DAreRodz commented on assets/js/interactivity/vdom.js:
```diff
@@ -60,11 +78,17 @@ export function toVdom( root ) {
 					let val = attributes[ i ].value;
 					try {
-						val = JSON.parse( val );
+						const [ json, ns ] = nsPathParser( val );
+						val = JSON.parse( json );
+						if ( ns && val ) val[ $$namespace ] = ns;
```
This is kind of a hack. Directives that hold a string value (a.k.a. path) already contain the namespace in its value, but for those that hold a JSON, I had no idea where to keep the namespace. E.g.:

```html
<div
  data-wp-on--click="custom-ns::actions.clickHandler"
  data-wp-context='custom-ns::{ "count": 0 }'
> ... </div>
```

As we parse the JSON string inside `toVdom()`, I ended up using a `Symbol()` to store the namespace directly in the resulting object.

@luisherranz, let me know what you think.

luisherranz replied:
Oh yeah, we need to store a namespace _per_ directive. That's right.

I'm not sure I like to expose a Symbol as an API: `[ context[ $$namespace ] || namespace ]: context`. It's weird and not easy to follow.

Maybe we need to rethink the directives structure to something more like this:

```html
<div data-wp-interactive='{ "namespace": "myPlugin" }'>
  <div
    data-wp-context='{ "isOpen": true }'
    data-wp-context--context2='otherPlugin::{ "isOpen": false }'
  >
    ...
  </div>
</div>
```

```js
// what directives("context", { directives }) receives
[
  {
    namespace: "myPlugin",
    suffix: "default",
    value: {
      isOpen: true,
    },
  },
  {
    namespace: "otherPlugin",
    suffix: "context2",
    value: {
      isOpen: false,
    },
  },
];
```

luisherranz replied:
And if we store namespaces inside the directives themselves, maybe there's no need to store them in the components.

DAreRodz replied:
That would make sense. Let's see how complex it becomes to consume a `directives` property with that format.

DAreRodz replied:
The suggested format for parsed directives was implemented in e4180a7ebf89456e1386e9253e7e2359ae8e36a8.

I also changed `evaluate()` to receive the whole directive definition instead of just the path.

DAreRodz commented on assets/js/interactivity/index.js:
```diff
@@ -1,13 +1,18 @@
+ */
 import registerDirectives from './directives';
 import { init } from './router';
 import { rawStore, afterLoads } from './store';
```
We import `rawStore` to pass it to `afterLoad` callbacks. This is not required anymore as you could use `store()` to get a reference to any store you need.

DAreRodz replied:
Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/826bb6e75fe2726fabf1ee2db7c012c8af61509a.

luisherranz replied:
I was thinking that `afterLoad` is not needed anymore. Is it?

DAreRodz replied:
Uhm. We added that API to initialize stuff that depended on the store's _readiness_. Do we have something in the new API that replaces that? 🤔

luisherranz replied:
> We added that API to initialize stuff that depended on the store's readiness

Hmm... after all the `store()` calls have been executed, you mean?

If we look at this: https://github.com/woocommerce/woocommerce-blocks/blob/interactivity-api-refactor-product-button-with-new-store-api/assets/js/atomic/blocks/product-elements/button/frontend.tsx#L128-L146

The need for `afterLoad` is not justified anymore. You can just do this and it works fine:

```js
const { state } = store("woo", {
  // ....
});

subscribe(() => {
  const cartData = select(storeKey).getCartData();
  const isResolutionFinished =
    select(storeKey).hasFinishedResolution("getCartData");
  if (isResolutionFinished) {
    state.cart = cartData;
  }
}, storeKey);

requestIdleCallback(() => {
  if (!state.hasCartLoaded) {
    select(storeKey).getCartData();
  }
});
```

DAreRodz replied:
Oh, right. 😄  The `afterLoad` API was intended to get the store reference and use it in the code you wanted to execute.

Now I see why it's not needed anymore. I'll remove anything related to the `afterLoad` callbacks.

luisherranz replied:
As you said, the only use case that I see is to have a way to wait until all the stores have been merged, so if you are accessing some part of the store that is not defined in your file, it's there when you access `state`.

But we don't have any use case for that now, so my vote goes for removing it, and we'll evaluate it once we see the need for it again.

DAreRodz replied:
Done in 35354eaf1bb55258a72d466c6d5dea0655cab3b0.

DAreRodz committed:
Do not pass `rawStore` to `afterLoad` callbacks

DAreRodz committed:
Simplify `store()` a bit

DAreRodz committed:
Implement `privateStore()`

luisherranz committed:
Sync context directive with Gutenberg

luisherranz committed:
Refactor scope and extract getters per scope

luisherranz commented:
I've extracted the getters so different instances of them run "in the scope of each component". It's a bit like hooks, but automatic 😄

I've also refactored the `scope` a little bit. Now it makes more sense. You also need to pull `interactivity-api-refactor-product-button-with-new-store-api`.

I'll keep thinking if `scope` needs to be a stack, because now it needs to be set up inside the `computeds`, because they can run "outside" of the compontent scope when any of their dependencies is invalidated. At this point, I don't think it needs to be a stack because our directives don't stack, but if we don't make it a stack, there's no way to reset it properly either. So it'd be more a stack that consists of the same `scope` over and over to know when is the last and be able to properly reset it 😄

DAreRodz committed:
Add namespace to getters and actions

DAreRodz committed:
Remove current privateStore implementation

DAreRodz committed:
Remove `afterLoad` option from `store`

DAreRodz committed:
Use same proxy handlers for ns, getters and actions

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -52,26 +52,32 @@ const parseInitialState = () => {
+
+		// Check if the proxy is the store root and no prop with that name
+		// exist. In that case, return an empty object for the prop requested.
+		if ( receiver === stores.get( ns ) && ! ( key in target ) ) {
```
Great trick!

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -93,14 +99,10 @@ const stateHandlers = {
 
-const actionHandlers = {
-	get: ( target, key, receiver ) => {
 		const result = Reflect.get( target, key, receiver );
```
`const result = Reflect.get( target, key, receiver );`

Note: This could also be a getter if for some reason this runs outside of a scope. I can't think of any case where a getter might run outside of a scope and still need access to the default namespace, but maybe we find it in the future.

So there's nothing to do, this is just a note in case we find that case in the future.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,240 @@
+		if ( getter ) {
+			const scope = getScope();
+			if ( scope ) {
+				scope.getters = scope.getters || new Map();
```
Maybe we should move this to an external Map as well: scope -> getters, to avoid adding non-user-facing stuff to the scope.

DAreRodz replied:
Currently, the scope is not publicly accessible. Developers use `getElement` and `getContext`, which return only a subset of the current scope. But we can move `getters` outside and use a `WeakMap` to link them to the scope related. I think it makes sense, anyway, so I'll make this change. 👍

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,240 @@
+						} )
+					);
+				}
+				return scope.getters.get( getter ).value;
```
Note: We are going to need a way to access the computed when we optimize `wp-text` and `wp-bind` to pass signals down. It doesn't have to be a nice API, so maybe something as simple as `getComputed(obj, "prop");`

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,261 @@
+export const getElementRef = () => currentScope.ref.current;
+
+export const getScope = () => currentScope;
+export const setScope = ( scope ) => {
+	currentScope = scope;
+};
+export const resetScope = () => {
+	currentScope = null;
+};
+
+export const setNamespace = ( namespace ) => {
+	currentNamespace = namespace;
+};
+export const resetNamespace = () => {
+	currentNamespace = null;
+};
```
One of the next steps should be to turn these into stacks.

DAreRodz replied:
Done in 96631ce688118e399f249826a568e26819e53f74.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,240 @@
+					return await iterate( gen.next( res ) );
+				};
+				try {
+					return iterate( gen.next() );
```
I didn't include `setScope` here because previously this was not an stack and the scope is correct here, but maybe we should do it now. The same for `setNamespace` (which if we don't, would be missing on the first part of the generator).

```suggestion
					setScope( scope );
					setNamespace( proxyToNs.get( receiver ) );
					return iterate( gen.next() );
```

DAreRodz replied:
Fixed in 39cdbe357b2f18325299bb8c036dc4e27d8f5ee3.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,240 @@
+					resetNamespace();
+					if ( iteration.done ) return iteration.value;
+					const res = await iteration.value;
+					setNamespace( proxyToNs.get( receiver ) );
```
Minimal optimization: the `receive` never changes so we can get the namespace only once.

DAreRodz replied:
Done in 39cdbe357b2f18325299bb8c036dc4e27d8f5ee3.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,240 @@
+		// Check if the property is a function.
+		if ( typeof result === 'function' ) {
+			const scope = getScope();
+			return ( ...args ) => {
+				setNamespace( proxyToNs.get( receiver ) );
+				setScope( scope );
+				result( ...args );
+				resetScope();
+				resetNamespace();
+			};
```
Oh, I forgot that functions can return 😄 

```suggestion
			return ( ...args ) => {
				setNamespace( proxyToNs.get( receiver ) );
				setScope( scope );
				const ret = result( ...args );
				resetScope();
				resetNamespace();
				return ret;
			};
```

DAreRodz replied:
Oh, right! Fixed in 95378841448d28206058a944da98180a4a039536.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,240 @@
+		}
+
+		// Check if the property is an object.
+		if ( isObject( result ) ) return proxify( result, ns );
```
Note: This is reusing the `isObject` from `deepMerge` which excludes arrays, although arrays could potentially contain functions/generators. For now, I don't think we need to support that as I can't think of any case where someone would like to have actions inside an array.

So there's nothing to do, this is just a note in case we find that case in the future.

DAreRodz replied:
Yup, that's the reason I opted to use `isObject`. If we find we need to support actions inside arrays, we can change it in the future. 🙂

luisherranz reviewed (commented):
I did a small review of your last commit and left some notes, but this is quite exciting 🙂👏

Amazing work!!

DAreRodz committed:
Set scope inside `evaluate`

DAreRodz committed:
Refactor proxy handlers

DAreRodz committed:
Improve types a bit

DAreRodz committed:
Catch errors in async actions

DAreRodz committed:
Implement stacks for scopes and namespaces

DAreRodz committed:
Implement `getElement`

DAreRodz committed:
Change directives object structure

DAreRodz committed:
Remove unnecessary import

DAreRodz committed:
Implement private stores

DAreRodz committed:
Return value from sync actions

luisherranz commented on assets/js/interactivity/vdom.js:
```diff
@@ -19,6 +28,12 @@ const directiveParser = new RegExp(
 	'i' // Case insensitive.
 );
 
+const nsPathRegExp = /^([\w-]+)::(.+)$/;
```
This would benefit from a comment similar to the one on the directive parser regexp.

DAreRodz replied:
Fixed by @luisherranz in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/f1e0cdff72f9055cff0ae01dedfe88f6ba7b5ded

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,299 @@
+ */
+import { stores } from './store';
+
+/** @typedef {import('preact').VNode} VNode */
+/** @typedef {typeof context} Context */
+/** @typedef {ReturnType<typeof getEvaluate>} Evaluate */
+
+/**
+ * @typedef {Object} DirectiveCallbackParams Callback parameters.
+ * @property {Object}   directives Object map with the defined directives of the element being evaluated.
+ * @property {Object}   props      Props present in the current element.
+ * @property {VNode}    element    Virtual node representing the original element.
+ * @property {Context}  context    The inherited context.
+ * @property {Evaluate} evaluate   Function that resolves a given path to a value either in the store or the context.
+ */
+
+/**
+ * @callback DirectiveCallback Callback that runs the directive logic.
+ * @param {DirectiveCallbackParams} params Callback parameters.
+ */
+
+/**
+ * @typedef DirectiveOptions Options object.
+ * @property {number} [priority=10] Value that specifies the priority to
+ *                                  evaluate directives of this type. Lower
+ *                                  numbers correspond with earlier execution.
+ *                                  Default is `10`.
+ */
```
Can we replace all this with proper TypeScript types, or is it needed for the docs?

DAreRodz replied:
I think we can. Let me try.

luisherranz replied:
But have you confirmed that WordPress docs can use TypeScript instead of JSDocs?

DAreRodz replied:
I haven't yet, but I can see `.ts` files with mixed doc formats... 👀

DAreRodz replied:
@luisherranz, I just checked this. We can use TS docs without any issues. Apart from that, we would need to include these tags in the `README.md` file of the `@wordpress/interactivity` package, where we want the API reference to be added.

```md
<!-- START TOKEN(Autogenerated API docs) -->
<!-- END TOKEN(Autogenerated API docs) -->
```

Then, simply run `npm run  docs:api-ref` and, voilà! ✨


<details>
<summary>I did a quick test, and this is (part of) the result.</summary>
<br>

---

### store

Extends the Interactivity API global store with the passed properties.

These props typically consist of `state`, which is reactive, and other properties like `selectors`, `actions`, `effects`, etc. which can store callbacks and derived state. These props can then be referenced by any directive to make the HTML interactive.

_Usage_

```js
const { state } = store( 'counter', {
	state: {
		value: 0,
	},
	actions: {
		increment: () => {
			state.counter.value += 1;
		},
	},
} );
```

The code from the example above allows blocks to subscribe and interact with
the store by using directives in the HTML, e.g.:

```html
<div data-wp-interactive='{ "namespace": "counter" }'>
	<button data-wp-text="state.value" data-wp-on--click="actions.increment">
		0
	</button>
</div>
```

_Parameters_

-   _namespace_ `string`: The namespace the store will use.
-   _properties_ `S`: Data to add to the store with the passsed namespace.
-   _properties.state_ `S[ 'state' ]`: Reactive part of the store.
-   _options_ `StoreOptions`: Options passed to the `store` call.
-   _options.lock_ `StoreOptions[ 'lock' ]`: Whether this store should be private.
</details>

luisherranz replied:
Cool 🙂 

cc: @juanmaguitar

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,299 @@
+	return Object.freeze( {
+		ref: ref.current,
+		state: state.current,
+		props: deepImmutable( props ),
```
Note: maybe one day we can remove the deep immutable logic from the production version.

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,299 @@
+};
+
+export const getElement = () => {
+	if ( ! getScope() ) {
+		throw Error(
+			'Cannot call `getElement()` outside getters and actions used by directives.'
+		);
+	}
```
Note: We should start wrapping these with `if (SCRIPT_DEBUG)` in Gutenberg.

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,299 @@
+	const scope = useRef( {} ).current;
+	scope.context = useContext( context );
+	scope.ref = previousScope.ref || useRef( null );
+	scope.state = previousScope.state || useRef( deepSignal( {} ) );
```
Is there any benefit on storing the whole object instead of the `current` prop? The reference should be stable.

```suggestion
	scope.state = previousScope.state || useRef( deepSignal( {} ) ).current;
```

DAreRodz replied:
I know―I just followed the same logic as `scope.ref`. The `state.current` value is returned by `getElement()` as `state`. But I can change it if you want. 🙂

luisherranz replied:
I think it's slightly better, yes. We can leave `ref` as it is because that is updated without being reactive.

DAreRodz replied:
Done in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/48adf8388568fc38ac29ffc15618e226cbad6e06.

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,299 @@
+	scope.context = useContext( context );
+	scope.ref = previousScope.ref || useRef( null );
+	scope.state = previousScope.state || useRef( deepSignal( {} ) );
+	scope.props = element?.props || originalProps;
```
When is element not defined?

DAreRodz replied:
Oh, right. The `element` prop is always defined. I'll update that.

PS: I've noticed that we're adding the props of the previous element instance (we're cloning the element for each priority level). I'm not sure if that could cause problems accessing newly added properties, but just to prevent any potential issue, we could clone `element` before adding `props` to the scope. 🤔

luisherranz replied:
Yeah, I think we need to review all the props and element logic at some point. It's a bit convoluted.

DAreRodz replied:
Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/0662935488af86c982fbe16001a201b887ca8887.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,281 @@
+	try {
+		const initialState = JSON.parse( storeTag.textContent );
+		if ( isObject( initialState ) ) return initialState;
+		throw Error( 'Parsed state is not an object' );
```
Note: we should wrap this with `SCRIPT_DEBUG` when in Gutenberg.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,281 @@
+
+export function store< S extends object = {} >(
+	namespace: string,
+	storePart?: DeepPartial< S >,
```
Should we add this here or should we let people use it themselves? 🤔

DAreRodz replied:
Uhm, I just spent a short time thinking about this, and I'm not sure. It is better if we help developers not deal with that.

The issue is whether we want devs to get types inferred when they pass a _store part_ to `store()`. E.g., let's consider the following store:

```ts
interface MyStore {
  state: {
    prop1: string;
    prop2: string;
  }
}
```

It would be ideal if we have the arguments typed in the following call automatically:

```ts
const { state } = store< MyStore >( 'myStore', {
  state: {
    prop1: 123, // ERROR: This cannot be a number!
    prop3: 456, // ERROR: This prop does not exist!
  },
});
```

If we don't handle that internally, developers would have to do something like this:

```ts
const { state } = store< MyStore >(
  'myStore',
  {
    state: { prop1: '123' },
  } as DeepPartial< MyStore > // Just to infer types correctly.
);
```

Or perhaps they could write their own typed objects...


```ts
interface StorePart {
  state: {
    prop1: string;
  }
}

const storePart: StorePart  = {
  state: { prop1: '123' },
};

const { state } = store< MyStore >('myStore', storePart);
```

Do you have any other idea in mind, @luisherranz?

luisherranz replied:
My idea was to encourage the use of separate types in different calls to `store()`: one with the definition with the partial types, and one with the `state` and `actions` destructuring with the whole types.

It's a bit more complex, but the types are perfect. I explained it here in the `Objects returned by store() are typed
` section:

- https://github.com/WordPress/gutenberg/discussions/53586#:~:text=Objects%20returned%20by%20store()%20are%20typed

_By perfect, I mean that if some store props are missing, TS will complain._

My fear is that by adding `DeepPartial` inside the types of `store`, there's no way to make the types "perfect".

luisherranz replied:
I'm not 100% sure that's the ideal scenario, though. We need to do more testing.

DAreRodz replied:
Aha, thanks for the reminder; I forgot that part of the proposal. 😅 

Then, it makes sense to try without `DeepPartial` and see how it goes.

luisherranz replied:
I think it's safer for now, yes 🙂

DAreRodz replied:
Removed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/2750a072597dc434ca28199330b6c7ce4d22a6ef.

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,281 @@
+		} else {
+			const storeLock = storeLocks.get( namespace );
+			const isLockValid =
+				lock === universalUnlock ||
+				( lock !== true && lock === storeLock );
```
Note: Maybe we could store `false` as the lock for public stores to simplify this logic a bit. But it's not important at this moment.

DAreRodz replied:
Hey, that's how it already works! 😄

https://github.com/woocommerce/woocommerce-blocks/blob/ebe6f622e53f06abff104e0bf894e840e49bbe69/assets/js/interactivity/store.ts#L233

The thing is that stores locked with `lock: true` can only be unlocked with the `universalLock` (i.e., passing `true` should not unlock it!). That's why that case is covered.

luisherranz replied:
Oh, ok. Thanks 🙂

luisherranz commented on assets/js/interactivity/directives.js:
```diff
@@ -28,79 +42,96 @@ export default () => {
+					}, [ newContext, inheritedValue ] );
+
+					return (
+						<Provider value={ currentValue.current }>
+							{ toReturn }
+						</Provider>
```
Why are you wrapping providers with the one another instead of reusing the same one?

DAreRodz replied:
Completely right. I'll update that part.

DAreRodz replied:
Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/f0508c973da57f1d3f061e161865354b088c1aaa.

luisherranz commented on assets/js/interactivity/directives.js:
```diff
@@ -28,79 +42,96 @@ export default () => {
+		watch.forEach( ( entry ) => {
+			useSignalEffect( async () => {
+				const result = evaluate( entry );
+				return await result;
```
What is the logic to use an async function here?

DAreRodz replied:
I'm not sure, to be honest. I simply preserved the current logic, and I don't remember who wrote it. 🤔

PS: I think it was you when you replaced `effect` with `watch`. 😄 c467f35fde5abaa5c16b3d5696f10dd12cddc003

luisherranz replied:
Oh, haha, sorry. I don't know what I was thinking about (generators?), but I think it's wrong. We should return the result of evalute directly.

luisherranz replied:
Done in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/9ae9002d0a08f83491d9cff40289326c9569ec3c.

luisherranz committed:
Minor optimizations and improved comments

luisherranz reviewed (commented):
Some of the new directives are pretty neat 😄

```js
directive("init", ({ directives: { init }, evaluate }) => {
  init.forEach((entry) => {
    useEffect(() => evaluate(entry), []);
  });
});
```

I did a first review and everything seems to be working fine. Great work, David! 🎉 🎉

However, I'm starting to have second thoughts about doing all this without tests. Maybe we should open a PR in Gutenberg to make sure that it works with our current tests, and instead of merging that, come back here to keep testing it with the Woo blocks.

What do you think?

DAreRodz commented:
> Maybe we should open a PR in Gutenberg to make sure that it works with our current tests, and instead of merging that, come back here to keep testing it with the Woo blocks.
> 
> What do you think?

Good idea! 👍 

We'll have to do it anyway soon, so it will be well-spent time.

Also, I have already needed to do manual testing for everything. Automated testing would be greatly appreciated. 🙏

luisherranz commented on assets/js/interactivity/directives.js:
```diff
@@ -28,79 +42,90 @@ export default () => {
+	// data-wc-watch--[name]
+	directive( 'watch', ( { directives: { watch }, evaluate } ) => {
+		watch.forEach( ( entry ) => {
+			useSignalEffect( async () => {
+				const result = evaluate( entry );
+				return await result;
 			} );
```
```suggestion
			useSignalEffect( () => evaluate( entry ) );
```

luisherranz committed:
Don't use async inside `data-wp-watch`

luisherranz commented:
Great. I have added a list of tasks to the Gutenberg's Tracking Issue:

- https://github.com/WordPress/gutenberg/issues/53740

DAreRodz committed:
Use a single Provider in context directive

DAreRodz committed:
Remove DeepPartial type

DAreRodz committed:
Do not check if element exists

DAreRodz committed:
Add the `current` prop of state inside the scope

DAreRodz committed:
Move getters outside scope

luisherranz commented on assets/js/interactivity/hooks.tsx:
```diff
@@ -0,0 +1,298 @@
+
+// Resolve the path to some property of the store object.
+const resolve = ( path, namespace ) => {
+	let current = {
+		...stores.get( namespace ),
+		context: getScope().context[ namespace ],
+	};
```
We need to add `elementState` here as well. We probably need to find a better name than Element State.

DAreRodz committed:
Fix wc-key assignment

DAreRodz committed:
Merge branch 'trunk' into interactivity-api-new-store-api

DAreRodz committed:
Fix missing `navigate` in directives

luisherranz committed:
Fix namespace not being picked in the same element

luisherranz committed:
Merge branch 'trunk' into interactivity-api-new-store-api

luisherranz commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,284 @@
+		}
+
+		const target = stores.get( namespace );
+		deepMerge( target, block );
+		deepMerge( target.state, state );
```
We shouldn't merge the proxies, but the raw store objects. We can store them on another map.

DAreRodz replied:
Fixed in https://github.com/woocommerce/woocommerce-blocks/commit/e1c10075261498c662b53d34d0fd77c070be33e9. I've also tested in the Gutenberg branch to ensure tests keep passing.

luisherranz replied:
Thanks, David!

DAreRodz committed:
Deep merge raw stores instead of proxied ones

DAreRodz committed:
Fix namespace assignment

DAreRodz commented:
I've found another problem with namespaces while working in Gutenberg. When a previous namespace was assigned, and a directive was defined in an element with a new one, it received the previous one instead.

Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/8e605e62d9f50cb9b34030ac9fe867d62d10a03b.

cc: @luisherranz

luisherranz commented:
Good catch 👍

DAreRodz committed:
Allow forward slashes in namespaces

DAreRodz commented:
Another thing I've found is that we are using forward slashes in namespaces, e.g.

https://github.com/woocommerce/woocommerce-blocks/blob/200b38303496cffc6052260f782070e5e77666ed/src/BlockTypes/ProductButton.php#L168

But we didn't allow that character in namespaces for directives defining it explicitly, so the following didn't work:

```html
<div data-wp-watch="woocommerce/product-button::callbacks.start animation"></div>
```

Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/5d91a56a0d4134258407f14946688e35f2f1f2b6.

DAreRodz committed:
Merge branch 'trunk' into interactivity-api-new-store-api

Aljullu commented on assets/js/interactivity/directives.js:
```diff
@@ -155,56 +144,155 @@ export default () => {
 				// Prefetch the page if it is in the directive options.
 				if ( link?.prefetch ) {
-					prefetch( href );
+					// prefetch( href );
```
Why was this commented out? Is it intended to leave it like this?

DAreRodz replied:
Uh, funny. 😄 That's actually the code we have in Gutenberg right now:

https://github.com/WordPress/gutenberg/blob/05432dd7c26f4c282b8750141a86a4b154704972/packages/interactivity/src/directives.js#L292-L294

I'd leave it as it is for now, although that should change eventually.

Aljullu commented on assets/js/interactivity/store.ts:
```diff
@@ -0,0 +1,285 @@
+	resetNamespace,
+} from './hooks';
+
+const isObject = ( item: unknown ): boolean =>
+	!! item && typeof item === 'object' && ! Array.isArray( item );
```
I don't know if you can have `@woocommerce` dependencies in this file. But in case you can, we have an [`isObject` function](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/types/type-guards/object.ts#L7-L15) in `@woocommerce/types` which might be useful.

DAreRodz replied:
I would prefer not to depend on other files not part of the Interactivity API runtime. 🤔

Aljullu reviewed (commented):
I have reviewed the code and tested this PR via #11558 and couldn't spot anything off. However, as I mentioned in https://github.com/woocommerce/woocommerce-blocks/pull/11558#pullrequestreview-1724892417, I still feel I'm missing a lot of context around how things work with the Interactivity API, so I don't think I'm entitled to approve or reject this PR. I will let the others take a look.

Nevertheless, I left a couple of comments in the code. They are both minor suggestions/questions, so feel free to disregard them.

luisherranz committed:
Migration of Product Collection and Product Button blocks to the new `store()` API (#11558)

* Refactor Product Button with new store() API

* Use `wc_initial_state` in Product Button

* Fix namespace

* Remove unnecessary state

* Test namespaces in directive paths

* Add test context with namespace

* Simplify woo-test context

* Move addToCart and animations to a file

* Do not pass `rawStore` to `afterLoad` callbacks

* Move callbacks and actions back to the main file

Because the animation was broken.

* Remove selectors in favor of state

* Use default ns in `getContext` for state and actions

* Remove `afterLoad` callback

* Remove unnecessary ns

* Fix getContext in add-to-cart

* Replace namespace and delete unnecessary store

* Pass context types only once

* Use an alternative for requestIdleCallback

* Add previous react code for notices

* Add namespace to Product Collection block

* Replace getTextButton with getButtonText

* Add block name to the ProductCollection namespace

* fix style HTML code

* Remove circular deps error on the Interactivity API

* Product Gallery block: Migrate to new Interactivity API store (#11721)

* Migrate Product Gallery block to new Interactivity API store

* Fix some references

* Add missing data-wc-interactive

* Fix an additional namespace

* Remove unnecessary click handler

* Dialog working

* Refactor action names

* Reindex PHP array

There was some missing indexes, which turned the array into an object in JS.

* Remove unused event handlers

* Move next/previous logic to external function

* Move StorePart util to the types folder

* Rename namespace to `woocommerce/product-gallery`

* Undo product collection namespace renaming

* Remove unnecessary namespace

* Don't hide the large image on page load

* Minor refactorings

* Fix eslint error

* Fix php cs errors with spacing and double arrows alignment

* Disable no-use-before-define rule for eslint

* Disable @typescript-eslint/ban-types rule for eslint

* Fix parsed context error in e2e tests

* Fix context parser for Thumbnail image

* Move store to the top of the frontend file

* Add interactivity api utils to the @woocommerce/utils alias

* Replace deprecated event attribute

---------

Co-authored-by: Luis Herranz <luisherranz@gmail.com>
Co-authored-by: David Arenas <david.arenas@automattic.com>
Co-authored-by: roykho <roykho77@gmail.com>

---------

Co-authored-by: David Arenas <david.arenas@automattic.com>
Co-authored-by: Luigi Teschio <gigitux@gmail.com>
Co-authored-by: Alexandre Lara <allexandrelara@gmail.com>
Co-authored-by: roykho <roykho77@gmail.com>

DAreRodz committed:
Merge branch 'trunk' into interactivity-api-new-store-api

DAreRodz committed:
Merge branch 'trunk' into interactivity-api-new-store-api

roykho committed:
Fix error when closing product gallery dialog with keyboard escape key

luisherranz reviewed (approved):
@gigitux can you apply the changes that you did in https://github.com/woocommerce/woocommerce-blocks/pull/11858 also here please?

Are the e2e tests already fixed in `trunk`? Who's taking care of that? I'd like to merge this PR to continue the migration.

---

As there are conflicts with https://github.com/woocommerce/woocommerce-blocks/pull/11858, I'm going to anticipately approve this PR, but the fix in that PR needs to be applied here as well before merging.

cc: @DAreRodz @sunyatasattva

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into interactivity-api-new-store-api

gigitux commented:
> @gigitux can you apply the changes that you did in #11858 also here please?

Done!

Currently the E2E tests pass on `trunk`. So, if something doesn't pass, maybe it could be worth investigating it!

luisherranz committed:
Merge commit 'a94a442d9a4cf40cb3cc68e08e172ccba1304bd2' into interactivity-api-new-store-api

luisherranz commented:
@gigitux merging `trunk` with this branch didn't apply the fix of https://github.com/woocommerce/woocommerce-blocks/pull/11858, it deleted `wc_initial_state`: https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/09821c98f13925486ddd9bca99db0d2fc22916e6

<img width="1086" alt="Screenshot 2023-11-21 at 11 24 09" src="https://github.com/woocommerce/woocommerce-blocks/assets/3305402/1873678d-edb7-46f5-99e9-38dbd9c093ed">

gigitux committed:
use wc_initial_state instead of wc_store
